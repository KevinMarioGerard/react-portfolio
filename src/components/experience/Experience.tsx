import * as React from 'react';
import ExperienceContext, { SelectedState } from './ExperienceContext';
import { ExperienceTabs, ExperienceTab } from './ExperienceTabs';
import ExperienceItem from './ExperienceItem';

type ExperienceContentProps = {
    /**
     * id mapped to experience tab.
     */
    id: string;
};

const ExperienceContent: React.FC<ExperienceContentProps> = ({ children, id }) => {
    const { selected } = React.useContext(ExperienceContext);

    return (
        <>
            {id === selected ? (
                <div className="experience__content__container" id={id}>
                    {children}
                </div>
            ) : null}
        </>
    );
};

ExperienceContent.displayName = 'Experience.Content';

type Statics = {
    Content: typeof ExperienceContent;

    Tabs: typeof ExperienceTabs;

    Tab: typeof ExperienceTab;

    Item: typeof ExperienceItem;
};

const Experience: React.FC & Statics = ({ children }) => {
    const [selected, setSelected] = React.useState<SelectedState>(null);

    return (
        <div className="experience__container">
            <ExperienceContext.Provider value={{ selected, setSelected }}>
                {children}
            </ExperienceContext.Provider>
        </div>
    );
};

Experience.displayName = 'Experience';

Experience.Content = ExperienceContent;
Experience.Item = ExperienceItem;
Experience.Tabs = ExperienceTabs;
Experience.Tab = ExperienceTab;

export default Experience;
