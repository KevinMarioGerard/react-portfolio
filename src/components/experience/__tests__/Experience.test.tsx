import { render, fireEvent, getByText } from '@testing-library/react';
import Experience from '../Experience';
import * as React from 'react';

const TestExperience = () => (
    <Experience>
        <Experience.Tabs initial="test1">
            <Experience.Tab id="test1" label="Tab 1" />
            <Experience.Tab id="test2" label="Tab 2" />
            <Experience.Tab id="test3" label="Tab 3" />
        </Experience.Tabs>
        <Experience.Content id="test1">
            <Experience.Item
                title="Associate Software Engineer"
                company="Test 1"
                startDate="July 2019"
                endDate="Present"
                description={['description line 1', 'description line 2']}
            />
        </Experience.Content>
        <Experience.Content id="test2">
            <Experience.Item
                title="Software Engineer"
                company="Test 2"
                startDate="July 2019"
                endDate="Present"
                description={['description line 1', 'description line 2']}
            />
        </Experience.Content>
        <Experience.Content id="test3">
            <Experience.Item
                title="Full Stack Developer"
                company="Test 3"
                startDate="July 2019"
                endDate="Present"
                description={['description line 1', 'description line 2']}
            />
        </Experience.Content>
    </Experience>
);

describe('Experience Component tests', () => {
    it('should render without errors', () => {
        expect(() => {
            render(<TestExperience />);
        }).not.toThrow();
    });

    it('should match snapshot', () => {
        const { container } = render(<TestExperience />);

        expect(container).toMatchSnapshot();
    });

    it('should update content on click', () => {
        const { container } = render(<TestExperience />);

        fireEvent.click(getByText(container, 'Tab 3'));

        expect(container).toMatchSnapshot();
    });
});
