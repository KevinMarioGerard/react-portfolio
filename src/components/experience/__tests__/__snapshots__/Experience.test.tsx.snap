// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experience Component tests should match snapshot 1`] = `
<div>
  <div
    class="experience__container"
  >
    <div
      class="experience__tabs"
    >
      <div
        class="experience__tabs__container"
      >
        <div
          class="experience__tab experience__tab--selected"
          id="test1"
        >
          <span>
            Tab 1
          </span>
        </div>
        <div
          class="experience__tab "
          id="test2"
        >
          <span>
            Tab 2
          </span>
        </div>
        <div
          class="experience__tab "
          id="test3"
        >
          <span>
            Tab 3
          </span>
        </div>
      </div>
    </div>
    <div
      class="experience__content__container"
      id="test1"
    >
      <div
        class="experience__item__container"
      >
        <h3
          class="experience__item__header"
        >
          <span
            class="experience__item__title"
          >
            Associate Software Engineer
          </span>
          <span
            class="experience__item__company"
          >
             @ 
            <span>
              Test 1
            </span>
          </span>
        </h3>
        <span
          class="experience__item__duration"
        >
          July 2019
           - 
          Present
        </span>
        <div
          class="experience__item__description"
        >
          <ul>
            <li>
              description line 1
            </li>
            <li>
              description line 2
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Experience Component tests should update content on click 1`] = `
<div>
  <div
    class="experience__container"
  >
    <div
      class="experience__tabs"
    >
      <div
        class="experience__tabs__container"
      >
        <div
          class="experience__tab "
          id="test1"
        >
          <span>
            Tab 1
          </span>
        </div>
        <div
          class="experience__tab "
          id="test2"
        >
          <span>
            Tab 2
          </span>
        </div>
        <div
          class="experience__tab experience__tab--selected"
          id="test3"
        >
          <span>
            Tab 3
          </span>
        </div>
      </div>
    </div>
    <div
      class="experience__content__container"
      id="test3"
    >
      <div
        class="experience__item__container"
      >
        <h3
          class="experience__item__header"
        >
          <span
            class="experience__item__title"
          >
            Full Stack Developer
          </span>
          <span
            class="experience__item__company"
          >
             @ 
            <span>
              Test 3
            </span>
          </span>
        </h3>
        <span
          class="experience__item__duration"
        >
          July 2019
           - 
          Present
        </span>
        <div
          class="experience__item__description"
        >
          <ul>
            <li>
              description line 1
            </li>
            <li>
              description line 2
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
